### PR TITLE
added cache for DidResolver

### PR DESCRIPTION
**Pull request by Cxxshyy**
Added in a cache that would check if the did being looked at has been cached and already has a resolution result value prepared and if not to use the resolve function to get the resolution result, then place the value in cache with the did as the key.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -35,34 +38,44 @@ npm link ../path/to/where/you/cloned/dwn-sdk-js
...
+  ></script>
   ```
 
 - **ESM Bundle**
   ```html
-  <script type="text/javascript" src="node_modules/dwn-sdk/dist/bundles/bundle.esm.js"></script>
+  <script
+    type="text/javascript"
+    src="node_modules/dwn-sdk/dist/bundles/bundle.esm.js"
+  ></script>
```
Comment (thehenrytsai) [inline README.md:72]: Why the change to double quotes and formatting? Consider undo.

Comment (Cxxshyy) [reply README.md:72]: not to sure, i ran n lint:fix and this was the output

File: src/did/did-resolver.ts [hunk 2]
```diff
@@ -1,12 +1,14 @@
...
 
 /**
  * TODO: add docs, Issue #72 https://github.com/TBD54566975/dwn-sdk-js/issues/72
  */
 export class DidResolver {
   didResolvers: Map<string, DidMethodResolver>;
 
+  memoryCache = new MemoryCache(86400);//Time to live in seconds is 24 hours
   // TODO: add DIDCache to constructor method signature, Issue #62 https://github.com/TBD54566975/dwn-sdk-js/issues/62
   constructor(resolvers: DidMethodResolver[]) {
```
Comment (thehenrytsai) [inline src/did/did-resolver.ts:13]: Pass in the cache through constructor as an optional parameter such that it can be changed and mocked. e.g.

```suggestion
  constructor(resolvers: DidMethodResolver[], cache?: Cache) {
```

Also the TODO should be removed since this PR addresses the issue.

File: src/did/did-resolver.ts [hunk 3]
```diff
@@ -36,7 +38,14 @@ export class DidResolver {
       throw new Error(`${didMethod} DID method not supported`);
     }
 
-    const resolutionResult = await didResolver.resolve(did);
+    const resolutionResult = await this.memoryCache.has(did) ? await this.memoryCache.get(did): await didResolver.resolve(did);
+    const cache = await this.memoryCache.has(did);
+    if (cache){
+      await this.memoryCache.set(did,resolutionResult);
```
Comment (thehenrytsai) [inline src/did/did-resolver.ts:44]: Looks like a bug? If result is in cache, no need to store it in cache again, you meant to do `!cache`?

File: src/did/did-resolver.ts [hunk 4]
```diff
@@ -36,7 +38,14 @@ export class DidResolver {
       throw new Error(`${didMethod} DID method not supported`);
     }
 
-    const resolutionResult = await didResolver.resolve(did);
+    const resolutionResult = await this.memoryCache.has(did) ? await this.memoryCache.get(did): await didResolver.resolve(did);
+    const cache = await this.memoryCache.has(did);
```
Comment (thehenrytsai) [inline src/did/did-resolver.ts:42]: rename `cache` to `isCached`

File: src/did/did-resolver.ts [hunk 5]
```diff
@@ -36,7 +38,14 @@ export class DidResolver {
       throw new Error(`${didMethod} DID method not supported`);
     }
 
-    const resolutionResult = await didResolver.resolve(did);
+    const resolutionResult = await this.memoryCache.has(did) ? await this.memoryCache.get(did): await didResolver.resolve(did);
```
Comment (thehenrytsai) [inline src/did/did-resolver.ts:41]: There is no need to introduce `.has()`, `.get()` will return `undefined` if not in cache, you can use that as your condition.

Comment (Cxxshyy) [reply src/did/did-resolver.ts:41]: i did use that first but thought the .has would be more clean

File: src/did/did-resolver.ts [hunk 6]
```diff
@@ -1,12 +1,14 @@
...
 import { Did } from './did';
+import { MemoryCache } from '../utils/memory-cache';
 
 /**
  * TODO: add docs, Issue #72 https://github.com/TBD54566975/dwn-sdk-js/issues/72
  */
 export class DidResolver {
   didResolvers: Map<string, DidMethodResolver>;
 
+  memoryCache = new MemoryCache(86400);//Time to live in seconds is 24 hours
```
Comment (thehenrytsai) [inline src/did/did-resolver.ts:11]: 24 hours is rather aggressive as default, change to say 10 minutes.

File: src/did/did-resolver.ts [hunk 7]
```diff
@@ -36,7 +38,14 @@ export class DidResolver {
       throw new Error(`${didMethod} DID method not supported`);
     }
 
-    const resolutionResult = await didResolver.resolve(did);
+    const resolutionResult = await this.memoryCache.has(did) ? await this.memoryCache.get(did): await didResolver.resolve(did);
+    const cache = await this.memoryCache.has(did);
+    if (cache){
```
Comment (thehenrytsai) [inline src/did/did-resolver.ts:43]: Will need to add tests to cover the code added.

**Review by thehenrytsai — CHANGES_REQUESTED** [review]
@Cxxshyy thanks again for your 3rd PR! 🙏 🏆 🎸 
This one requires more changes though!

File: src/did/did-resolver.ts (hunk 5 above)
Comment (thehenrytsai) [reply src/did/did-resolver.ts:41]: Probably logically more clean, but "dirtier" in practice as it requires two round trips to the cache.

**Comment by Cxxshyy** [comment]
@thehenrytsai yeah thanks i'm still learning typescript/javascript and appreciate the guidance🤌

File: src/did/did-resolver.ts (hunk 2 above)
Comment (Cxxshyy) [reply src/did/did-resolver.ts:13]: sorry dont understand this, though i did add the cache as an optional parameter however what do you mean by dependency on interface

Comment (thehenrytsai) [reply src/did/did-resolver.ts]: > what do you mean by dependency on interface

The DidResolver should have no reference/knowledge of "memory cache" except for the construction in constructor.

> also what operations are to be done on the cache passed in?

Not sure what you mean.

Comment (Cxxshyy) [reply src/did/did-resolver.ts]: ok thanks

Comment (Cxxshyy) [reply src/did/did-resolver.ts]: the thing still confusing me is the fact that we have the memoryCache being instantiated but then if we are passing in a cache as an optional parameter to allow for ease of testing what operations does the cache passed in do? For example it cant instantiate the memoryCache class as it requires a parameter of TTL(time to live) meaning in cant modify that class unless im missing something, so essentially what do i do with the cache parameter passed in.

Comment (thehenrytsai) [reply src/did/did-resolver.ts]: You will only instantiate the memory cache if no cache is passed in assuming cache is an optional parameter, in which case you should simply instantiate the cache with a default timeout as suggested. The design of cache as an "optional" parameter is simply a convenience, if the developer doesn't like to use memory cache or the default timeout set for memory cache, fine, pass in their own.

One could argue for making the cache always required to be passed in, I am okay with that, not too opinionated on that, but dependency injection of the cache is a must.

Comment (Cxxshyy) [reply src/did/did-resolver.ts]: ok thanks that makes sense but cba you also check the issues comments as there is a new problem affecting testing of the code

Comment (Cxxshyy) [reply src/did/did-resolver.ts]: can

Comment (thehenrytsai) [reply src/did/did-resolver.ts]: Answered in issue #62

Comment (Cxxshyy) [reply src/did/did-resolver.ts]: i have another question so right now i have modified my code to have a boolean arg which will allow for injection of a new cache and if its false will use memorycache instantiating and youre able to pass in a timeout else default value is 600.

Now the question is if the type of cache can be modified then how will that affect the functions currently being used for example .get() cannot be used if the type of cache is Cache but can be used if its MemoryCache?

Comment (Cxxshyy) [reply src/did/did-resolver.ts]: unless im being confused and cache:Cache is only meant to be used for testing?

**Comment by thehenrytsai** [comment]
@Cxxshyy, checkout PR https://github.com/TBD54566975/dwn-sdk-js/pull/133, you can follow the pattern for adding the cache!

File: src/did/did-resolver.ts (hunk 2 above)
Comment (thehenrytsai) [reply src/did/did-resolver.ts]: Cache is the interface. So as long as object of type Cache is passed in, the resolver can use it the same way regardless of implementation.

**Comment by Cxxshyy** [comment]
hi @thehenrytsai im having trouble identifying the error 
image

**Comment by Cxxshyy** [comment]
image

**Comment by thehenrytsai** [comment]
`resolvers` is being passed in as `undefined` in  `constructor(resolvers: DidMethodResolver[], cache? :Cache)`.

It appears you got rid of a block of code when you pulled and merged main into your branch:

https://github.com/TBD54566975/dwn-sdk-js/blob/main/src/did/did-resolver.ts#L13

**Comment by Cxxshyy** [comment]
thanks another issue 
image
for some reason im getting an error related to a file i never modified

**Comment by Cxxshyy** [comment]
i also tried to write a test if you can also check it thanks its in did-resolver.spec.ts

**Comment by thehenrytsai** [comment]
I'd suggest that given there is not a lot of changes, and I see a some incorrect/unintended changes probably due branch merges.

1. stash/save your changes
2. pull and revert back to latest clean state
3. run tests to make sure everything passes
4. apply/add your changes on top, if tests starts to break, you know is your changes for sure.

**Comment by Cxxshyy** [comment]
Ok all the code ran fine but now when i try using (this.cache = cache || new MemoryCache(600);) it comes up saying (Type 'Cache | MemoryCache' is not assignable to type 'Cache'). Even though it should be allowed

**Comment by Cxxshyy** [comment]
nevermind got it

**Comment by Cxxshyy** [comment]
if possible can close this PR and use the new one i created @thehenrytsai

**Comment by thehenrytsai** [comment]
Happy to! Sorry for the slow response, have been out of town last week but back now.
